Completion Suggester: Add doc-values based payload support

areek opened the pull request:
Note: the PR is against `feature/completion_suggester_v2` branch

Followup of https://github.com/elastic/elasticsearch/pull/13576#issuecomment-140695698. The PR enables returning any doc values enabled fields as suggestion payload.

One can specify fields to be returned as part of the suggestion payload at query time, using the 
`payload` option:
```bash
POST music/_suggest 
{
  "song-suggest" : {
    "prefix" : "nev",
    "completion" : {
      "field" : "song_suggest"
      "payload" : [ "title" ]
    }
  }
}
```
The specified payload field values will be returned as part of the suggestion `payload`:
```bash
{
  "song-suggest": [
    {
      "text": "nev",
      "offset": 0,
      "length": 4,
      "options" : [ {
        "text" : "Nirvana",
        "score" : 34.0,
        "payload" : {
          "title" : [ "Nevermind" ]
        }
      } ]
    }
  ]
}
```

Now suggestion `payload` are not part of the completion (FST) index, as was the case before. Payload fields are fetched for the top N completions per shard hence specifying payload fields will incur additional search performance hit.

mikemccand commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestParser.java:
```diff
@@ -158,6 +160,14 @@ public CompletionSuggestParser(CompletionSuggester completionSuggester) {
                 }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                if ("payload".equals(fieldName)) {
+                    while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
```
Maybe a bit better error checking here?  E.g. if you see a START_OBJECT token, that's not allowed right?

areek replied:
Thanks for the pointer, I have made it stricter and also added tests to ensure it errors out on malformed input

mikemccand commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java:
```diff
@@ -67,18 +71,36 @@ public SuggestContextParser getContextParser() {
+                    payload = new LinkedHashMap<>(payloadFields.size());
+                    for (String field : payloadFields) {
+                        MappedFieldType fieldType = suggestionContext.getMapperService().smartNameFieldType(field);
+                        if (fieldType != null) {
```
Hmm, if `fieldType` is null it means the use asked for a non-existent payloads field?  Can we throw an exc (earlier) so user realizes this?

areek replied:
Changed to throw an exception when we encounter non-existent payload fields

mikemccand commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionContext.java:
```diff
@@ -34,16 +37,19 @@
     }
 
-    CompletionFieldMapper.CompletionFieldType fieldType() {
+    CompletionFieldMapper.CompletionFieldType getFieldType() {
```
++ for the traditional naming :)

mikemccand commented on docs/reference/search/suggesters/completion-suggest.asciidoc:
```diff
@@ -144,17 +146,71 @@ POST music/_suggest?pretty
 The `text` field uses the `input` of your indexed suggestion.
 
-The basic completion suggester query supports the following two parameters:
+Suggestions are document oriented, you can specify any doc values
```
Won't non doc-values fields work too (inverted into field data), since you are using the field data APIs?  Not sure we should allow that though ...

areek replied:
Yes, you are right. It will work with non doc-values fields too. I have updated the docs to reflect that, but not sure if we should just encourage using doc values enabled fields only?

mikemccand commented on docs/reference/search/suggesters/completion-suggest.asciidoc:
```diff
@@ -144,17 +146,71 @@ POST music/_suggest?pretty
 
 `field`:: The name of the field on which to run the query (required).
 `size`::  The number of suggestions to return (defaults to `5`).
+`payload`::  The name of the fields to be returned as payload (defaults to no fields).
```
Maybe "field or fields array" to make it clear it can be a single field or multiple ones?

areek replied:
Changed

mikemccand commented on docs/reference/search/suggesters/completion-suggest.asciidoc:
```diff
@@ -144,17 +146,71 @@ POST music/_suggest?pretty
-NOTE: It will be possible to return the associated documents
-with each suggestion in the future (TODO).
+NOTE: Specifying `payload` fields will incur additional search performance
+hit. The `payload` fields are retrieved for top suggestions at the shard
```
Maybe say "retrieved eagerly (single pass)" to make it clear there is not a separate fetch phase?

mikemccand replied:
Maybe also say fields can be numeric, date, etc. too?  (Since this is a new improvement vs. current suggesters)

areek replied:
I have made the NOTE clearer about how the payload fields are retrieved, and added the support for various field types in the intro paragraph illustrating how to use `payload`

mikemccand commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestion.java:
```diff
@@ -54,6 +54,48 @@ public CompletionSuggestion(String name, int size) {
+        if (toReduce.size() == 1) {
+            return toReduce.get(0);
+        } else {
+            OptionPriorityQueue priorityQueue = new OptionPriorityQueue(size);
```
Maybe some docs explaining what `OptionPriorityQueue` is doing?  I.e. we are combining multiple results from N shards on the coordinating node right?

areek replied:
Added docs to clarify

mikemccand commented:
This looks great, thanks @areek, I left a few small comments.

Net/net this is more functionality vs. current suggesters, since you can retrieve more than one field as the payload (vs only one payload per suggestion today), and they can be typed fields (numbers, dates, etc.) too.

The code is also quite simpler than supporting the 2nd fetch phase ...

rmuir commented:
Wait we are using docvalues as stored fields now? I do not think this is something we should do.

s1monw commented:
> Wait we are using docvalues as stored fields now? I do not think this is something we should do.

I think for the suggest usecase that's a good alternative to putting it into the FST. We will deprecate this option as well once we have the second roundtrip query_then_fetch for this too but that will go into 3.x

areek commented:
Thanks @mikemccand for the review, addressed all your feedback. I think the docs can be improved to encourage users to use doc values enabled fields for payload? Maybe @clintongormley has some thoughts?

@rmuir this is added to make the new implementation have feature parity with the current suggester and as @s1monw mentioned will go away in 3.x, where documents will be fetched in the 2nd phase.

clintongormley commented:
@areek I like it

mikemccand commented:
Thanks @areek, LGTM.
